--- cpyrit/cpyrit.h	2008-12-29 18:17:35.000000000 +0100
+++ cpyrit/cpyrit.h	2008-12-29 18:23:13.000000000 +0100
@@ -21,11 +21,12 @@
 #ifndef CPYRIT
 #define CPYRIT
 
-#include <python2.5/Python.h>
+#include <string.h>
 #include <stdint.h>
 #include <pthread.h>
 #include <openssl/hmac.h>
 #include <openssl/sha.h>
+#include <python2.6/Python.h>
 #ifdef HAVE_CUDA
     #include <cuda_runtime.h>
 #endif
